lib/os/queued_operation.c

```diff
@@ -0,0 +1,531 @@
+
+	if (res >= 0) {
+		if (from == ST_STARTING) {
+			trivial_start_and_unlock(mgr, key);
+			return;
+		}
+
+		/* Came from STOPPING or RESETTING so set to OFF.  If
+		 * the underlying onoff service is in an error state
+		 * that will be discovered when it is next used.
+		 */
+		mgr->state = ST_OFF;
+
+		k_spin_unlock(&mgr->lock, key);
+		return;
+	}
+
+	/* On transition failure mark service failed.  All unstarted
+	 * operations are unlinked and completed as a service failure.
+	 */
+	sys_slist_t ops = mgr->operations;
+
+	sys_slist_init(&mgr->operations);
+	mgr->state = ST_ERROR;
+
+	k_spin_unlock(&mgr->lock, key);
+
+	struct queued_operation *op;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&ops, op, node) {
+		finalize_and_notify(mgr, op, -ENODEV);
+	}
+}
+
+static void start_callback(struct onoff_manager *oomgr,
+			   struct onoff_client *cli,
+			   u32_t state,
+			   int res)
+{
+	struct queued_operation_manager *mgr
+		= CONTAINER_OF(cli, struct queued_operation_manager, onoff_client);
+
+	settle_onoff(mgr, ST_STARTING, res);
+}
+
+static void stop_and_unlock(struct queued_operation_manager *mgr,
+			    k_spinlock_key_t key)
+{
+	__ASSERT_NO_MSG(mgr->state == ST_IDLE);
+
+	if (mgr->onoff == NULL) {
+		mgr->state = ST_OFF;
+		k_spin_unlock(&mgr->lock, key);
+		return;
+	}
+
+	mgr->state = ST_STOPPING;
+
+	struct onoff_manager *onoff = mgr->onoff;
+
+	k_spin_unlock(&mgr->lock, key);
+
+	settle_onoff(mgr, ST_STOPPING, onoff_release(onoff));
+}
+
+/* Enters while locked, and looks for a new operation to initiate.
+ * The lock is released to process new operations, then retaken after
+ * process() returns.
+ */
+static void select_next_and_unlock(struct queued_operation_manager *mgr,
+				   k_spinlock_key_t key)
+{
+	bool loop = false;
+
+	do {
+		struct queued_operation *op = NULL;
+		sys_snode_t *node = sys_slist_get(&mgr->operations);
+		u32_t state = mgr->state;
+
+		__ASSERT_NO_MSG((state == ST_IDLE)
+				|| (state == ST_FINALIZING));
+		loop = false;
+		if (node) {
+			op = CONTAINER_OF(node, struct queued_operation, node);
+			mgr->state = ST_NOTIFYING;
+			mgr->current = op;
+
+			k_spin_unlock(&mgr->lock, key);
+
+			/* Notify the service, then check everything again
+			 * because the operation might have completed or the
+			 * queue might have changed while we were unlocked.
+			 */
+			mgr->vtable->process(mgr, op);
+
+			/* Update the state to one of IDLE, PROCESSING, or
+			 * leave it FINALIZING; loop if something needs to be
+			 * done.
+			 */
+			key = k_spin_lock(&mgr->lock);
+
+			state = mgr->state;
+
+			/* If an operation finalized during notification we
+			 * need to reselect because finalization couldn't do
+			 * that, otherwise it's still running.
+			 */
+			loop = (state == ST_FINALIZING);
+			if (!loop) {
+				__ASSERT_NO_MSG(state == ST_NOTIFYING);
+				mgr->state = ST_PROCESSING;
+			}
+		} else {
+			__ASSERT_NO_MSG(state == ST_FINALIZING);
+			mgr->state = ST_IDLE;
+			mgr->current = op;
```

Comment:
what about setting it to null and moving `op` variable to `if(node)` section?

Reply:
Could `stop_and_unlock(mgr, key);` be called here? Is there a case when `node` is not null and state is IDLE?